webroot/js/components/chat/chat-input.js

```diff
@@ -0,0 +1,281 @@
+import { generatePlaceholderText, getCaretPosition } from '../../utils/chat.js';
+import { getLocalStorage, setLocalStorage } from '../../utils/helpers.js';
+import { URL_CUSTOM_EMOJIS, KEY_CHAT_FIRST_MESSAGE_SENT } from '../../utils/constants.js';
+
+
+export default class ChatInput extends Component {
+  constructor(props, context) {
+    super(props, context);
+    this.formMessageInput = createRef();
+    this.emojiPickerButton = createRef();
+
+    this.messageCharCount = 0;
+		this.maxMessageLength = 500;
+    this.maxMessageBuffer = 20;
+
+    this.emojiPicker = null;
+
+    this.prepNewLine = false;
+
+    this.state = {
+      inputHTML: '',
+      inputWarning: '',
+      hasSentFirstChatMessage: getLocalStorage(KEY_CHAT_FIRST_MESSAGE_SENT),
+    };
+
+    this.handleEmojiButtonClick = this.handleEmojiButtonClick.bind(this);
+    this.handleEmojiSelected = this.handleEmojiSelected.bind(this);
+    this.getCustomEmojis = this.getCustomEmojis.bind(this);
+
+    this.handleMessageInputKeydown = this.handleMessageInputKeydown.bind(this);
+    this.handleMessageInputKeyup = this.handleMessageInputKeyup.bind(this);
+    this.handleMessageInputBlur = this.handleMessageInputBlur.bind(this);
+    this.handleSubmitChatButton = this.handleSubmitChatButton.bind(this);
+    this.handlePaste = this.handlePaste.bind(this);
+
+    this.handleContentEditableChange = this.handleContentEditableChange.bind(this);
+  }
+
+  componentDidMount() {
+    this.getCustomEmojis();
+  }
+
+  getCustomEmojis() {
+    fetch(URL_CUSTOM_EMOJIS)
+      .then(response => {
+        if (!response.ok) {
+          throw new Error(`Network response was not ok ${response.ok}`);
+        }
+        return response.json();
+      })
+      .then(json => {
+        this.emojiPicker = new EmojiButton({
+          zIndex: 100,
+          theme: 'dark',
+          custom: json,
+          initialCategory: 'custom',
+          showPreview: false,
+          emojiSize: '30px',
+          // emojisPerRow: 6,
+          position: 'right-start',
+          strategy: 'absolute',
+        });
+        this.emojiPicker.on('emoji', emoji => {
+          this.handleEmojiSelected(emoji);
+        });
+      })
+      .catch(error => {
+        // this.handleNetworkingError(`Emoji Fetch: ${error}`);
+      });
+  }
+
+  handleEmojiButtonClick() {
+    if (this.emojiPicker) {
+      this.emojiPicker.togglePicker(this.emojiPickerButton.current);
+    }
+  }
+
+  handleEmojiSelected(emoji) {
+    const { inputHTML } = this.state;
+    let content = '';
+    if (emoji.url) {
+      const url = location.protocol + "//" + location.host + "/" + emoji.url;
+      const name = url.split('\\').pop().split('/').pop();
+      content = "<img class=\"emoji\" alt=\"" + name + "\" src=\"" + url + "\"/>";
+    } else {
+      content = emoji.emoji;
+    }
+
+    this.setState({
+      inputHTML: inputHTML + content,
+    });
+  }
+
+  // autocomplete user names
+	autoCompleteNames() {
```

Comment:
nit: indents in this file generally seem a little off compared to the rest of the code

Reply:
weird, they looked fine in my editor, but re-tabbed some lines again just to make sure.